listeners.js

```diff
@@ -0,0 +1,179 @@
+      }),
+      // to find new epochs we only check groups we've accepted the invite to
+      paraMap((msg, cb) => {
+        pull(
+          ssb.box2.listGroupIds(),
+          pull.filter((groupId) => groupId === msg.value.content.recps[0]),
+          pull.take(1),
+          pull.collect((err, groupIds) => {
+            // prettier-ignore
+            if (err) return cb(clarify(err, "Error getting groups we're already in when looking for new epochs"))
+            cb(null, groupIds.length ? msg : null)
+          })
+        )
+      }, 4),
+      pull.filter(Boolean),
+      pull.drain(
+        (msg) => {
+          const groupId = msg.value.content.recps[0]
+
+          const secret = Buffer.from(msg.value.content.groupKey, 'base64')
+          ssb.box2.addGroupInfo(groupId, { key: secret }, (err) => {
+            // prettier-ignore
+            if (err && !isClosed) return onError(clarify(err, 'Cannot add new epoch key that we found'))
+
+            const newKeyPick = {
+              key: secret,
+              scheme: keySchemes.private_group,
+            }
+            // TODO: naively guessing that this is the latest key for now
+            ssb.box2.pickGroupWriteKey(groupId, newKeyPick, (err) => {
+              // prettier-ignore
+              if (err && !isClosed) return onError(clarify(err, 'Error switching to new epoch key that we found'))
+
+              ssb.db.reindexEncrypted((err) => {
+                // prettier-ignore
+                if (err && !isClosed) onError(clarify(err, 'Error reindexing after finding new epoch'))
+              })
+            })
+          })
+        },
+        (err) => {
+          // prettier-ignore
+          if (err && !isClosed) return onError(clarify(err, "Error finding new epochs we've been added to"))
+        }
+      )
+    )
+
+    // listen for new epochs and update groupInfo as required
+    pull(
+      ssb.db.query(
+        where(or(type('group/init'), type('group/add-member'))),
+        live({ old: true }),
+        toPullStream()
+      ),
+      pull.filter((msg) => isInitEpoch(msg) || isAddMember(msg)),
+      pull.map((msg) => msg.value.content.recps[0]),
+      pull.drain(
+        (groupId) => {
+          ssb.box2.getGroupInfo(groupId, (err, groupInfo) => {
+            // prettier-ignore
+            if (err && !isClosed) return onError(clarify(err, 'fatal error in live updating groupInfo'))
+            if (!groupInfo) return // group that we've not accepted an invite to yet
+            if (groupInfo.excluded) return // group where we were excluded
+
+            getPreferredEpoch(groupId, (err, epochInfo) => {
+              // prettier-ignore
+              if (err && !isClosed) return onError(clarify(err, 'fatal error getting preferred epoch'))
+              if (groupInfo.writeKey.key.equals(epochInfo.secret)) return
+
+              ssb.box2.addGroupInfo(
+                groupId,
+                { key: epochInfo.secret },
+                (err) => {
```

Comment:
I think this is unlikely to be needed here, we would need to have already added the epoch secret to figure out that the epoch is preferred i think

Reply:
but doesn't technically hurt and we'll be poking around here more later